###Coverage plots
setwd("~/Documents/P. Ovale Genomic Analysis")
library(ggplot2)
cov <- read.table("P. o. pop genomics data tables - Po coverage and depth.tsv", header = T, sep = "\t")
cov$group <- as.factor(cov$species)
#cov$format <- "format"
#boxplot(cov, col = c("skyblue3","red4","orange2"), outpch = , xlab = "Minimum depth of coverage",  ylab = "Proportion of genome covered", main = "Boxplot of sample genome coverage")
#map <- read.table("ov1r_mappingpercent_histo.txt", header = TRUE, sep = "\t")
#ov1cov <- c(0.9123149286, 0.9102402857, 0.8586938571, 0.8953662857, 0.9119336429, 0.8984735714, 0.8457744286, 0.8615220714, 0.8654318571, 0.8872014286, 0.9869772143, 0.9813826429, 0.9827180714, 0.7683915, 0.986713, 0.5125219286, 0.9787485, 0.984586, 0.9824175, 0.9827681429, 0.9815116429, 0.9851388571, 0.8326321429, 0.9814417857, 0.8875898571, 0.9796814286, 0.9812399286, 0.9821895714, 0.9832027143)
#hist(cov$, xlab = "Proportion of loci with at least 10X coverage", ylab = "Number of samples", main = "Histogram of genome coverage", col = c("skyblue3"), breaks = 10, xlim = c(0.5,1))
#hist(map$map, xlab = "% of reads mapped to reference genome", ylab = "Number of samples", main = "Histogram of % of reads mapped", col = c("skyblue3"), breaks = 10, xlim = c(0,100))        
#hist(map$depth, xlab = "Median depth of coverage (reads per site)", ylab = "Number of samples", main = "Histogram of median genome depth of coverage", col = c("skyblue3"), breaks = 15)

#ggplot(cov, aes(x=Coverage10X, fill = group)) + geom_histogram(binwidth = 0.05, fill = "skyblue3") + theme_bw() + scale_fill_manual(values = c("curtisi" = "skyblue3","mixed" = "mediumpurple3","wallikeri" = "green4")) + labs(x="Proportion of genome with at least 10 reads mapped", y = "Frequency", fill="species") + ggtitle("Histogram of 10X genome coverage per sample") + theme(plot.title = element_text(hjust =0.5), ) + theme_bw()+ theme(axis.title = element_text(size = 18),axis.text= element_text(size = 12), title = element_text(size = 20), legend.text = element_text(size = 15))
ggplot(cov, aes(x=Coverage10X, fill = group)) + geom_histogram(binwidth = 0.05) + theme_bw() + scale_fill_manual(values = c("curtisi" = "skyblue3","mixed" = "mediumpurple3","wallikeri" = "green4")) + labs(x="Proportion of genome with >10 reads mapped", y = "Frequency",fill="Species") + ggtitle("Histogram of 10X genome coverage per sample") + theme(plot.title = element_text(hjust =0.5)) + theme_bw()+ theme(axis.title.y = element_text(size = 22),axis.title.x = element_text(size=22),title = element_text(size = 20), legend.text = element_text(size = 15),axis.text= element_text(size = 12))
ggplot(cov, aes(x=Mapping., fill = group)) + geom_histogram(binwidth = 0.01) + theme_bw() + scale_fill_manual(values = c("curtisi" = "skyblue3","mixed" = "mediumpurple3","wallikeri" = "green4")) + labs(x="% Mapping to P. ovale reference genome after competitive alignment", y = "Frequency",fill="Species") + ggtitle("Histogram of ovale Mapping Percentage") + theme(plot.title = element_text(hjust =0.5)) + theme_bw()+ theme(axis.title.y = element_text(size = 22),axis.title.x = element_text(size=15), title = element_text(size=20),axis.text= element_text(size = 12),legend.text = element_text(size = 15))
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     , hjust=0.45),axis.text= element_text(size = 20), title = element_text(size = 20), legend.title = element_text(size = 24),legend.text = element_text(size = 24))
